# Deep Learning: Image Classification

Zening Qu (zqu@uw.edu), Sean Yang

Last updated: May 11, 2020

## Introduction

People rarely train deep neural networks from scratch (with random initialization) these days. Instead, we often use some form of [transfer learning](http://cs231n.github.io/transfer-learning/), which uses a [pretrained network](https://keras.rstudio.com/articles/applications.html) to save training time and improve model prediction. In this assignment, we will explore and compare two widely-used transfer learning techniques: feature extraction and fine-tuning. 


**Dataset**

We will use the [Viziometrics dataset](https://canvas.uw.edu/files/59549223/download?download_frd=1) ([paper1](https://ieeexplore.ieee.org/abstract/document/7888968), [paper2](https://arxiv.org/abs/1908.07465)). The dataset contains 2k+ images from scientific publications that are hand-labeled as one of the five classes: equation, photo, scheme, table, and visualization. The train, validation, and test set has already been splitted for us. 


**Tasks**

We have two tasks. In part 1, "feature extraction", we will use a pretrained network, ResNet50, to extract features from the five classes of images. Using these extracted features and labels, we will train a neural network and make predictions. This approach is called "feature extraction" because we use a pretrained network to extract features and then use the features ***instead of*** our raw image data to train our model for prediction.

In part 2, "fine-tuning", we will append our own prediction layers to ResNet50; we will train this modified network on the raw images and the labels. This approach is called "fine tuning" because ResNet50 was pretrained on ImageNet but we are fine-tuning the model weights on our own dataset. We are not learning the weights of this complicated modified model from scratch; we start from the ResNet50 pretrained weights.

As you complete the two tasks you will notice that, if we use the same prediction layers, fine-tuning gives better validation accuracy than feature extraction. This is because with fine-tuning, there are a lot more layers at work during training. Because fine-tuning uses more layers, it will also take much longer time to train compared to feature extraction.


**Before You Start**

Please make sure you have [numpy](https://numpy.org/), [Keras](https://keras.io/) and one of its backend engines (e.g., [TensorFlow](https://www.tensorflow.org/install/)).


**What to Turn In**

We have provided some starter code. Each task you need to complete is marked with a red school bag 🎒and is worth a few points. Each task will either ask you to print something or report certain metrics. **Please submit your completed notebook with the code and the required output of each task. You will not receive full points if either code or output is missing.**

👉 **_Please start early. It takes time to train these networks._**

## 0. Load the Data (2 points)

🎒<font color='red'>(2 points)</font> Put the dataset in the same folder as this notebook, then run the following starter code. Because our training set contains a lot of images, we use [`ImageDataGenerator.flow_from_directory()`](https://keras.io/preprocessing/image/) to load images in batches. Please refer to [Vijayabhaskar's tutorial](https://medium.com/@vijayabhaskar96/tutorial-image-classification-with-keras-flow-from-directory-and-generators-95f75ebe5720) to understand the parameterization of the generators. This starter code should output something like:

```
Found xxx images belonging to 5 classes.
Found xxx images belonging to 5 classes.
Found xxx images belonging to 5 classes.
```

Please include this output in your submission.

In [1]:
import numpy as np
import keras
from keras.preprocessing.image import ImageDataGenerator
from keras.applications.resnet50 import ResNet50, decode_predictions
from keras import layers
from keras.models import Model, Sequential
from keras.callbacks.callbacks import ModelCheckpoint, EarlyStopping

Using TensorFlow backend.
C:\ProgramData\Anaconda3\lib\site-packages\tensorflow\python\framework\dtypes.py:516: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
C:\ProgramData\Anaconda3\lib\site-packages\tensorflow\python\framework\dtypes.py:517: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
C:\ProgramData\Anaconda3\lib\site-packages\tensorflow\python\framework\dtypes.py:518: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
C:\ProgramData\Anaconda3\lib\site-packages\tensorflow\python\framework\dtypes.py

In [2]:
# create train, validation, and test generators from our image directory

datagen = ImageDataGenerator()

train_generator = datagen.flow_from_directory(
  directory=r"./viziometrics/train/",
  target_size=(224, 224),
  color_mode="rgb",
  batch_size=32,
  class_mode="categorical",
  shuffle=False,
  seed=42
)

val_generator = datagen.flow_from_directory(
  directory=r"./viziometrics/val/",
  target_size=(224, 224),
  color_mode="rgb",
  batch_size=32,
  class_mode="categorical",
  shuffle=False,
  seed=42
)

test_generator = datagen.flow_from_directory(
  directory=r"./viziometrics/test/",
  target_size=(224, 224),
  color_mode="rgb",
  batch_size=1,
  class_mode=None,
  shuffle=False,
  seed=42
)

Found 2733 images belonging to 5 classes.
Found 1571 images belonging to 5 classes.
Found 1563 images belonging to 5 classes.


Found 2733 images belonging to 5 classes.  
Found 1571 images belonging to 5 classes.  
Found 1563 images belonging to 5 classes.

## 1. Feature Extraction (28 points)

We will use a pretrained network (ResNet50) to extract features from the images. Then we will train a not-so-deep neural network on the features (X) and the labels (y). Finally we will evaluate our network. Please read [Vikas' tutorial on feature extraction](https://www.learnopencv.com/keras-tutorial-transfer-learning-using-pre-trained-models/) for code examples on feature extraction; we will use the same pipeline here.

🎒<font color='red'>(2 points)</font> Run the following code to load a [ResNet50 network](https://keras.rstudio.com/reference/application_resnet50.html) pretrained on the [ImageNet](http://www.image-net.org/) dataset. Note that we say `include_top=False` to exclude the last (or top) fully-connected layer which is responsible for classifying the labels in ImageNet; we want to use the earlier layers to extract features from our own dataset. Use `model.summary()` to inspect the ResNet50 architecture. This function should print out all the layers - their names, types, output shapes etc. Please include the printout in your submission.

In [3]:
# download the pre-trained ResNet50 model
resnet = keras.applications.resnet.ResNet50(include_top=False, weights='imagenet', input_shape=(224,224,3))

# inspect the ResNet50 architecture
resnet.summary()

# observe that the last layer, conv5_block3_out (Activation), has output shape 7 * 7 * 2048
# let's save this dimension for later use
DIM = 7 * 7 * 2048


Model: "resnet50"
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
input_1 (InputLayer)            (None, 224, 224, 3)  0                                            
__________________________________________________________________________________________________
conv1_pad (ZeroPadding2D)       (None, 230, 230, 3)  0           input_1[0][0]                    
__________________________________________________________________________________________________
conv1_conv (Conv2D)             (None, 112, 112, 64) 9472        conv1_pad[0][0]                  
__________________________________________________________________________________________________
conv1_bn (BatchNormalization)   (None, 112, 112, 64) 256         conv1_conv[0][0]                 
__________________________________________________________________________________________

🎒<font color='red'>(4 points)</font> Use ResNet50 to extract features from the images. Print the shape of the extracted features.

👉 We use [`model.predict()`](https://keras.io/models/sequential/#predict_generator) to extract features. This function will return the output of the last layer of ResNet50, namely, conv5_block3_out (Activation). These are the features captured by the pretrained network.

👉 This step is a bit slow. Setting `verbose=1` can make the wait a bit easier.

In [6]:
# # use ResNet50 to extract features from the images
train_features = resnet.predict(train_generator, verbose=1)

val_features = resnet.predict(val_generator, verbose=1)

test_features = resnet.predict(test_generator, verbose=1)

1563/1563 [==============================] - 798s 511ms/step


In [7]:
print(train_features.shape, val_features.shape, test_features.shape)

(2733, 7, 7, 2048) (1571, 7, 7, 2048) (1563, 7, 7, 2048)


In [137]:
test_features.shape

(1563, 7, 7, 2048)

hint: if you encounter the error   
'ImportError: Could not import PIL.Image'  
try 'pip3 install pillow' and restart the notebook kernel  
see: https://stackoverflow.com/questions/48225729/importerrorcould-not-import-pil-image-working-with-keras-ternsorflow/50775336

🎒<font color='red'>(2 points)</font> Observe that the features extracted by ResNet50 are 4D arrays. Following common practice, we want to reshape the features into 2D arrays, which will be the input X of our not-so-deep neural network classifier. Run the following code and print the shape of the 2D arrays.

In [8]:
# reshape the features to 2D arrays
train_X = train_features.reshape((-1, DIM))
val_X = val_features.reshape((-1, DIM))
test_X = test_features.reshape((-1, DIM))

print(train_X.shape, val_X.shape, test_X.shape)

(2733, 100352) (1571, 100352) (1563, 100352)


🎒<font color='red'>(3 points)</font> Read the class labels from the generators. Print `test_labels`

In [9]:
# # read the class labels from the generators
train_labels = train_generator.classes
val_labels = val_generator.classes
test_labels = test_generator.classes
print(test_labels)

[0 0 0 ... 4 4 4]


🎒<font color='red'>(3 points)</font> Observe that the labels are not one-hot encoded. We need to one-hot encode these labels and they will be the y that our model will predict.

In [10]:
# # get one-hot encoding of labels
def get_one_hot(labels, nb_classes):
    res = np.eye(nb_classes)[np.array(labels).reshape(-1)]
    return res.reshape(list(labels.shape)+[nb_classes])

In [11]:
# # use get_one_hot()
NUM_CLASSES = 5
train_y = get_one_hot(train_labels,NUM_CLASSES)
val_y = get_one_hot(val_labels,NUM_CLASSES)
test_y = get_one_hot(test_labels,NUM_CLASSES)

🎒<font color='red'>(4 points)</font> Define neural network that contains a few layers. There is a three-layer example network below. Please modify this network - you can add layers, remove layers, or change some of the parameters. Please print out your model architecture using `model.summary()`.

* Tried the following layers, but the best accuracy was 0.64863, lower than the example model.  
model.add(layers.Dense(64, activation = "relu", input_shape=(DIM,)))   
model.add(layers.Dropout(0.5, noise_shape=None, seed=None))  
model.add(layers.Dense(NUM_CLASSES, activation = "sigmoid"))  
* Tried the following layers, but the best accuracy was 0.82050, lower than the example model.  
model.add(layers.Dense(1024, activation = "relu", input_shape=(DIM,)))   
model.add(layers.Dropout(0.5, noise_shape=None, seed=None))  
model.add(layers.Dense(NUM_CLASSES, activation = "softmax"))  
* Tried the following layers, but the best accuracy was 0.8237, lower than the example model.   
model.add(layers.Dense(1024, activation = "relu", input_shape=(DIM,)))     
model.add(layers.Dropout(0.1, noise_shape=None, seed=None))    
model.add(layers.Dense(NUM_CLASSES, activation = "softmax"))  
* Tried the following layers, but the best accuracy was 0.81477, lower than the example model.   
model.add(layers.Dense(768, activation = "relu", input_shape=(DIM,)))   
model.add(layers.Dropout(0.1, noise_shape=None, seed=None))  
model.add(layers.Dense(NUM_CLASSES, activation = "softmax"))  
* Tried the following layers, but the best accuracy was 0.7638, lower than the example model.   
model.add(layers.Dense(256, activation = "relu", input_shape=(DIM,)))   
model.add(layers.Dropout(0.5, noise_shape=None, seed=None))  
model.add(layers.Dense(NUM_CLASSES, activation = "softmax"))  
* so I use the example model

In [89]:
#sample model
model = Sequential()

model.add(layers.Dense(512, activation = "relu", input_shape=(DIM,))) 
model.add(layers.Dropout(0.2, noise_shape=None, seed=None))
model.add(layers.Dense(NUM_CLASSES, activation = "softmax"))

model.summary()

Model: "sequential_26"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_50 (Dense)             (None, 512)               51380736  
_________________________________________________________________
dropout_26 (Dropout)         (None, 512)               0         
_________________________________________________________________
dense_51 (Dense)             (None, 5)                 2565      
Total params: 51,383,301
Trainable params: 51,383,301
Non-trainable params: 0
_________________________________________________________________


🎒<font color='red'>(2 points)</font> Compile your model using [`model.compile()`](https://keras.io/models/model/#compile). Choose `sgd` optimizer and `categorical_crossentropy` loss. Set `metrics = ["accuracy"]`

In [90]:
# compile the model before you can train it
model.compile(optimizer=keras.optimizers.sgd(), loss='categorical_crossentropy',metrics=['accuracy'])

🎒<font color='red'>(7 points)</font> Train the model you defined for about 10 epochs, or until the validation accuracy reaches 0.8 (whichever comes earlier). Note that the model should be trained using `train_X`, `train_y`, `val_X`, and `val_y`. `train_X` and `val_X` are our extracted features; `train_y` and `val_y` are our one-hot-encoded labels. Do not use `train_generator` or `val_generator` to train this model because those generators represent our raw image data. We have already extracted features from the raw images and hence no longer need to use the images for training. Observe how the train accuracy and validation accuracy changes over epochs. Report (in a code comment) the best train and validation accuracy you've seen during training.

👉 Note that you can [check-point](https://keras.io/callbacks/#modelcheckpoint) model weights during training. After training, you can load back the weights of your best model.

In [91]:
# # save model weights while model is under training
es=EarlyStopping(monitor='val_accuracy',baseline=0.8,mode='max',verbose=1,patience=9)
cp=ModelCheckpoint('feature-extraction-weights.{epoch:02d}-{val_loss:.2f}.hdf5',
                   monitor='val_loss',
                   mode='min',
                   verbose=1,
                   save_best_only=True
                  ) 

# # train the model for about 10 epochs
# # each epoch takes about 20 seconds on my laptop
model.fit(
    x = train_X, 
    y = train_y, 
    epochs = 10,
    validation_data = (val_X,val_y),
    callbacks = [es,cp]
)

Train on 2733 samples, validate on 1571 samples
Epoch 1/10
2733/2733 [==============================] - 53s 19ms/step - loss: 2.1819 - accuracy: 0.8266 - val_loss: 1.3177 - val_accuracy: 0.5996

Epoch 00001: val_loss improved from inf to 1.31772, saving model to feature-extraction-weights.01-1.32.hdf5
Epoch 2/10
2733/2733 [==============================] - 34s 13ms/step - loss: 0.2570 - accuracy: 0.9104 - val_loss: 1.0056 - val_accuracy: 0.6518

Epoch 00002: val_loss improved from 1.31772 to 1.00560, saving model to feature-extraction-weights.02-1.01.hdf5
Epoch 3/10
2733/2733 [==============================] - 37s 13ms/step - loss: 0.1694 - accuracy: 0.9447 - val_loss: 0.7787 - val_accuracy: 0.7518

Epoch 00003: val_loss improved from 1.00560 to 0.77873, saving model to feature-extraction-weights.03-0.78.hdf5
Epoch 4/10
2733/2733 [==============================] - 38s 14ms/step - loss: 0.1405 - accuracy: 0.9510 - val_loss: 0.8072 - val_accuracy: 0.7422

Epoch 00004: val_loss did not im

🎒<font color='red'>(1 point)</font> Report the test accuracy of your best model.

In [92]:
# # load the weights of your best model (model.load_weights)
model.load_weights('feature-extraction-weights.05-0.59.hdf5')
# # measure test accuracy
scores = model.evaluate(test_X, test_y, verbose=0)
print("%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))

accuracy: 81.57%


The test accuracy is 81.57%.

## 2. Fine-tuning (10 points)

Now we will explore the second technique: fine-tuning. We will append our own classification layers to the ResNet50 layers and train the entire network (from images to labels). We will start over from the generators.

In [93]:
# create train, validation, and test generators from our image directory

datagen = ImageDataGenerator()

train_generator = datagen.flow_from_directory(
  directory=r"./viziometrics/train/",
  target_size=(224, 224),
  color_mode="rgb",
  batch_size=32,
  class_mode="categorical",
  shuffle=True,
  seed=42
)

val_generator = datagen.flow_from_directory(
  directory=r"./viziometrics/val/",
  target_size=(224, 224),
  color_mode="rgb",
  batch_size=32,
  class_mode="categorical",
  shuffle=True,
  seed=42
)

test_generator = datagen.flow_from_directory(
  directory=r"./viziometrics/test/",
  target_size=(224, 224),
  color_mode="rgb",
  batch_size=1,
  class_mode=None,
  shuffle=False,
  seed=42
)

Found 2733 images belonging to 5 classes.
Found 1571 images belonging to 5 classes.
Found 1563 images belonging to 5 classes.


🎒<font color='red'>(3 points)</font> Append classification layers to ResNet50. Please modify the example layers below to define your own classification layers. Print the network architecture of your own model.

In [94]:
x = resnet.output
x = layers.GlobalAveragePooling2D()(x)
x = layers.Dropout(0.1)(x)
#add a fully-connected layer
x = layers.Dense(256, activation='relu')(x)
predictions = layers.Dense(NUM_CLASSES, activation= 'softmax')(x)
model = Model(inputs = resnet.input, outputs = predictions)

model.summary()

Model: "model_1"
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
input_1 (InputLayer)            (None, 224, 224, 3)  0                                            
__________________________________________________________________________________________________
conv1_pad (ZeroPadding2D)       (None, 230, 230, 3)  0           input_1[0][0]                    
__________________________________________________________________________________________________
conv1_conv (Conv2D)             (None, 112, 112, 64) 9472        conv1_pad[0][0]                  
__________________________________________________________________________________________________
conv1_bn (BatchNormalization)   (None, 112, 112, 64) 256         conv1_conv[0][0]                 
____________________________________________________________________________________________

🎒<font color='red'>(1 point)</font> Compile your model using [`model.compile()`](https://keras.io/models/model/#compile). Choose `sgd` optimizer and `categorical_crossentropy` loss. Set `metrics = ["accuracy"]`

In [95]:
# # need to compile the model before training
model.compile(optimizer=keras.optimizers.sgd(), loss='categorical_crossentropy',metrics=['accuracy'])

🎒<font color='red'>(5 points)</font> Train the model you defined for about 10 epochs, or until the validation accuracy reaches 0.8 (whichever comes earlier). This model should be trained on the images and their labels, so we can use `train_generator` and `val_generator` to fit the model. Checkpoint model weights during training. Report (in a code comment) the best train and validation accuracy you've seen during training.

In [97]:
es=EarlyStopping(monitor='val_accuracy',baseline=0.8,mode='max',patience=5,verbose=1)
cp=ModelCheckpoint('feature-extraction-weights.{epoch:02d}-{val_loss:.2f}.hdf5',
                   monitor='val_loss',
                   mode='min',
                   verbose=1,
                   save_best_only=True
                  ) 

# # train the model for about 10 epochs
# # each epoch takes about 800 seconds on my laptop
STEP_SIZE_TRAIN=train_generator.n/train_generator.batch_size
STEP_SIZE_VAL=val_generator.n/val_generator.batch_size
model.fit_generator(generator=train_generator,
                    steps_per_epoch=STEP_SIZE_TRAIN,
                    validation_data=val_generator,
                    validation_steps=STEP_SIZE_VAL,
                    epochs=10,
                    callbacks=[es,cp]
)

Epoch 1/10
86/85 [==============================] - 5082s 59s/step - loss: 0.3149 - accuracy: 0.9034 - val_loss: 1.0023 - val_accuracy: 0.6620

Epoch 00001: val_loss improved from inf to 1.00229, saving model to feature-extraction-weights.01-1.00.hdf5
Epoch 2/10
86/85 [==============================] - 4909s 57s/step - loss: 0.0939 - accuracy: 0.9700 - val_loss: 0.4567 - val_accuracy: 0.7015

Epoch 00002: val_loss improved from 1.00229 to 0.45669, saving model to feature-extraction-weights.02-0.46.hdf5
Epoch 3/10
86/85 [==============================] - 4839s 56s/step - loss: 0.0449 - accuracy: 0.9872 - val_loss: 0.8428 - val_accuracy: 0.7377

Epoch 00003: val_loss did not improve from 0.45669
Epoch 4/10
86/85 [==============================] - 4894s 57s/step - loss: 0.0241 - accuracy: 0.9956 - val_loss: 0.1054 - val_accuracy: 0.8205

Epoch 00004: val_loss improved from 0.45669 to 0.10543, saving model to feature-extraction-weights.04-0.11.hdf5
Epoch 5/10
86/85 [=======================

🎒<font color='red'>(1 point)</font> Report the test accuracy of your best model.

I failed at using model.evaluate() to evaluate the test accuracy. So I changed the method to output the accuracy score.

In [160]:
# # load the weights of your best model (model.load_weights)
model.load_weights('feature-extraction-weights.07-0.01.hdf5')
# # measure test accuracy
# predict test labels
pred_test = model.predict(x=test_generator, steps=(test_generator.n/test_generator.batch_size))
pred_test = np.argmax(pred_test, axis=1)

from sklearn.metrics import classification_report, accuracy_score

#compare original test labels and predicted test labels
print("accuracy score is", accuracy_score(test_generator.classes,predIdxs))

print(classification_report(test_generator.classes, predIdxs))

accuracy score is 0.8202175303902751
              precision    recall  f1-score   support

           0       0.89      0.99      0.94       197
           1       0.86      0.95      0.90       363
           2       0.67      0.83      0.74       342
           3       0.79      0.95      0.87       217
           4       0.99      0.56      0.72       444

    accuracy                           0.82      1563
   macro avg       0.84      0.86      0.83      1563
weighted avg       0.85      0.82      0.81      1563

